src/java.base/share/classes/sun/nio/ch/ChannelInputStream.java

```diff
@@ -142,4 +142,75 @@ public void close() throws IOException {
         ch.close();
     }
 
+    private static final int TRANSFER_SIZE = 8192;
+
+    @Override
+    public long transferTo(OutputStream out) throws IOException {
+        if (out instanceof ChannelOutputStream cos) {
+            WritableByteChannel oc = cos.channel();
+
+            if (ch instanceof FileChannel fc) {
+                long pos = fc.position();
+                long size = fc.size();
+                long i = 0L;
+                for (long n = size - pos; i < n;
+                  i += fc.transferTo(pos + i, Long.MAX_VALUE, oc));
+                fc.position(size);
```

Comment:
The patch is improving but needs cleanup so that it is easy to maintain. I think I would move the I/O ops out of the update code and into the for statement/block. Also this will need the update to the channel position to be a finally block so that the it is adjusted when there are partial transfers.

Reply:
Moved I/O operations into the `for` statement by https://github.com/openjdk/jdk/pull/4263/commits/562b1023e62c4ff0a36e55ebc119cee6fb69809c.

Reply:
Correctly positioning channel in case of exception by https://github.com/openjdk/jdk/pull/4263/commits/ed49098a4712bf23cb6d218c27695717ba3312c2.

Reply:
Honestly, for me the code looks perfectly comprehensible now, so I actually do not know what to simplify any further (besides adjusting to personal code style likes or dislikes). If there actually *is* something hard to read or understand, please don't hesitate to express your detailed concerns and/or propose an actual change. I will be happy to adopt it in this PR.